trying to build for iOSSim fails

repo: https//github.com/miakoring/fosdem-scui-demo

@stackotter ’s idea of what it could be:
> i think it's because swift bundler assumes that ios-platform means single-architecture but then builds against the 'generic' simulator destination which is x86_64+arm64
so it builds the metadata as a single architecture static library and then tries to include it in the build which ends up actually being double architecture

```txt
⚠️ ld: ignoring file '/Users/miakoring/Documents/Fork/fosdem-scui-demo/.build/bundler/metadata/libmetadata.a': fat file missing arch 'x86_64', file has 'arm64'
⚠️ ld: Could not find or use auto-linked framework 'UIUtilities': framework 'UIUtilities' not found
⚠️ ld: Could not parse or use implicit file '/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/Frameworks/SwiftUICore.framework/SwiftUICore.tbd': cannot link directly with 'SwiftUICore' because product being built is not an allowed client of it
❌ Undefined symbols for architecture x86_64
❌   "__getSwiftBundlerMetadata", referenced from:
❌ ld: symbol(s) not found for architecture x86_64
❌ clang: error: linker command failed with exit code 1 (use -v to see invocation)
[SCUIDemo] Linking SCUIDemo
** BUILD FAILED **


The following build commands failed:
    Ld /Users/miakoring/Documents/Fork/fosdem-scui-demo/.build/arm64-apple-iphonesimulator/Build/Intermediates.noindex/SCUIDemo.build/Debug-iphonesimulator/SCUIDemo.build/Objects-normal/x86_64/Binary/SCUIDemo normal x86_64 (in target 'SCUIDemo' from project 'SCUIDemo')
    Building workspace fosdem-scui-demo with scheme SCUIDemo and configuration Debug
(2 failures)
error: Failed to run 'Failed to run xcodebuild.'

Caused by:
  Process didn't exit successfully: `/usr/bin/env xcodebuild -scheme SCUIDemo -configuration Debug -usePackageSupportBuiltinSCM -skipMacroValidation -derivedDataPath /Users/miakoring/Documents/Fork/fosdem-scui-demo/.build/arm64-apple-iphonesimulator -destination generic/platform=iOS\ Simulator OTHER_LDFLAGS=-lmetadata\ -L/Users/miakoring/Documents/Fork/fosdem-scui-demo/.build/bundler/metadata\ ${OTHER_LDFLAGS} OTHER_SWIFT_FLAGS=-DSWIFT_BUNDLER_METADATA\ ${OTHER_SWIFT_FLAGS} GCC_PREPROCESSOR_DEFINITIONS=SWIFT_BUNDLER_METADATA=1\ ${GCC_PREPROCESSOR_DEFINITIONS}` (exit status: 65)
```

--- a/Sources/SwiftBundler/Bundler/DeviceManager.swift
+++ b/Sources/SwiftBundler/Bundler/DeviceManager.swift
@@ -162,7 +162,7 @@ enum DeviceManager {
     try expect("}")
 
     guard let platform = dictionary["platform"] else {
-      throw failure("Missing platform")
+      throw failure("Missing platform (\(dictionary))")
     }
 
     let variant = dictionary["variant"]
@@ -182,7 +182,22 @@ enum DeviceManager {
     }
 
     guard let name = dictionary["name"] else {
-      throw failure("Missing name")
+      throw failure("Missing name '\(dictionary)'")
+    }
+
+    let architecture: BuildArchitecture
+    if parsedPlatform.supportedArchitectures.count == 1 {
+      architecture = parsedPlatform.supportedArchitectures[0]
+    } else {
+      guard let archString = dictionary["arch"] else {
+        throw failure("Missing arch (\(dictionary))")
+      }
+
+      guard let parsedArchitecture = BuildArchitecture(rawValue: archString) else {
+        throw failure("Unknown arch '\(archString)'")
+      }
+
+      architecture = parsedArchitecture
     }
 
     return Device(
@@ -190,7 +205,8 @@ enum DeviceManager {
       name: name,
       id: id,
       status: dictionary["error"].map(ConnectedDevice.Status.unavailable)
-        ?? .available
+        ?? .available,
+      architecture: architecture
     )
   }
 
@@ -200,7 +216,7 @@ enum DeviceManager {
   ) async throws(Error) -> Device {
     guard specifier != "host" else {
       if platform == nil || platform == HostPlatform.hostPlatform.platform {
-        return .host(HostPlatform.hostPlatform)
+        return .host(HostPlatform.hostPlatform, BuildArchitecture.host)
       } else {
         throw Error(.deviceNotFound(specifier: specifier, platform: platform))
       }

--- a/Sources/SwiftBundler/Bundler/HostPlatform.swift
+++ b/Sources/SwiftBundler/Bundler/HostPlatform.swift
@@ -1,5 +1,5 @@
 /// A supported Swift Bundler host platform.
-enum HostPlatform {
+enum HostPlatform: Sendable, Hashable {
   case macOS
   case linux
   case windows

--- a/Sources/SwiftBundler/Bundler/NonMacAppleOS.swift
+++ b/Sources/SwiftBundler/Bundler/NonMacAppleOS.swift
@@ -1,5 +1,5 @@
 /// A non-macOS Apple operating system.
-enum NonMacAppleOS: CaseIterable {
+enum NonMacAppleOS: Sendable, CaseIterable {
   case iOS
   case tvOS
   case visionOS

--- a/Sources/SwiftBundler/Bundler/NonMacApplePlatform.swift
+++ b/Sources/SwiftBundler/Bundler/NonMacApplePlatform.swift
@@ -2,7 +2,7 @@ import Foundation
 
 /// A non-macOS Apple platform. Used to model Apple platforms that can be
 /// connected to computers to become run destinations.
-enum NonMacApplePlatform: Equatable {
+enum NonMacApplePlatform: Sendable, Equatable {
   case physical(NonMacAppleOS)
   case simulator(NonMacAppleOS)
 

--- a/Sources/SwiftBundler/Bundler/ProjectBuilder.swift
+++ b/Sources/SwiftBundler/Bundler/ProjectBuilder.swift
@@ -423,7 +423,7 @@ enum ProjectBuilder {
         projectDirectory: scratchDirectory.builder,
         scratchDirectory: scratchDirectory.builder / ".build",
         configuration: .debug,
-        architectures: [.current],
+        architectures: [.host],
         platform: HostPlatform.hostPlatform.platform,
         additionalArguments: []
       ),

--- a/Sources/SwiftBundler/Bundler/Runner/ConnectedDevice.swift
+++ b/Sources/SwiftBundler/Bundler/Runner/ConnectedDevice.swift
@@ -1,9 +1,10 @@
 /// A connected device or simulator.
-struct ConnectedDevice: Equatable {
+struct ConnectedDevice: Sendable, Equatable {
   let platform: NonMacApplePlatform
   let name: String
   let id: String
   let status: Status
+  let architecture: BuildArchitecture
 
   enum Status: Equatable, CustomStringConvertible {
     case available

--- a/Sources/SwiftBundler/Bundler/Runner/Device.swift
+++ b/Sources/SwiftBundler/Bundler/Runner/Device.swift
@@ -2,25 +2,27 @@ import ArgumentParser
 import Foundation
 
 /// A device that can be used to run apps.
-enum Device: Equatable, CustomStringConvertible {
-  case host(HostPlatform)
+enum Device: Sendable, Equatable, CustomStringConvertible {
+  case host(HostPlatform, BuildArchitecture)
   /// Mac Catalyst isn't a host platform, because we don't run Swift Bundler under
   /// Mac Catalyst, so it can't live under the `.host` case. But for all intents
   /// and purposes, this `.macCatalyst` case functions very similarly to `.host`.
-  case macCatalyst
+  case macCatalyst(BuildArchitecture)
   case connected(ConnectedDevice)
 
   var description: String {
     switch self {
-      case .host(let platform):
-        return "\(platform.platform.name) host machine"
-      case .macCatalyst:
-        return "Mac Catalyst host machine"
+      case .host(let platform, let architecture):
+        return "\(platform.platform.name) host machine (arch: \(architecture.rawValue))"
+      case .macCatalyst(let architecture):
+        return "Mac Catalyst host machine (arch: \(architecture.rawValue))"
       case .connected(let device):
         return "\(device.name) (\(device.platform.platform), id: \(device.id))"
     }
   }
 
+  /// The device's id (as decided by the device's corresponding management tool
+  /// such as devicectl or adb).
   var id: String? {
     switch self {
       case .host, .macCatalyst:
@@ -30,9 +32,10 @@ enum Device: Equatable, CustomStringConvertible {
     }
   }
 
+  /// The device's platform.
   var platform: Platform {
     switch self {
-      case .host(let platform):
+      case .host(let platform, _):
         return platform.platform
       case .macCatalyst:
         return .macCatalyst
@@ -41,25 +44,37 @@ enum Device: Equatable, CustomStringConvertible {
     }
   }
 
+  /// The device's architecture.
+  var architecture: BuildArchitecture {
+    switch self {
+      case .host(_, let architecture), .macCatalyst(let architecture):
+        return architecture
+      case .connected(let device):
+        return device.architecture
+    }
+  }
+
   init(
     applePlatform platform: ApplePlatform,
     name: String,
     id: String,
-    status: ConnectedDevice.Status
+    status: ConnectedDevice.Status,
+    architecture: BuildArchitecture
   ) {
     switch platform.partitioned {
       case .macOS:
         // We assume that we only have one macOS destination so we ignore the
         // device id.
-        self = .host(.macOS)
+        self = .host(.macOS, architecture)
       case .macCatalyst:
-        self = .macCatalyst
+        self = .macCatalyst(architecture)
       case .other(let nonMacPlatform):
         self.init(
           nonMacApplePlatform: nonMacPlatform,
           name: name,
           id: id,
-          status: status
+          status: status,
+          architecture: architecture
         )
     }
   }
@@ -68,13 +83,15 @@ enum Device: Equatable, CustomStringConvertible {
     nonMacApplePlatform platform: NonMacApplePlatform,
     name: String,
     id: String,
-    status: ConnectedDevice.Status
+    status: ConnectedDevice.Status,
+    architecture: BuildArchitecture
   ) {
     let device = ConnectedDevice(
       platform: platform,
       name: name,
       id: id,
-      status: status
+      status: status,
+      architecture: architecture
     )
     self = .connected(device)
   }

--- a/Sources/SwiftBundler/Bundler/Runner/Runner.swift
+++ b/Sources/SwiftBundler/Bundler/Runner/Runner.swift
@@ -24,7 +24,7 @@ enum Runner {
     log.info("Running '\(bundlerOutput.bundle.lastPathComponent)'")
 
     switch device {
-      case .host(let platform):
+      case .host(let platform, let architecture):
         guard let bundlerOutput = RunnableBundlerOutputStructure(bundlerOutput) else {
           throw Error(.missingExecutable(device, bundlerOutput))
         }
@@ -37,12 +37,30 @@ enum Runner {
               environmentVariables: environmentVariables
             )
           case .linux:
+            guard architecture == .host else {
+              throw Error(cause: InvariantFailure(
+                """
+                Invariant failure: Runner received device with unsupported \
+                architecture '\(architecture.rawValue)' (host architecture: \
+                \(BuildArchitecture.host.rawValue))
+                """
+              ))
+            }
             try await runLinuxAppOnHost(
               bundlerOutput: bundlerOutput,
               arguments: arguments,
               environmentVariables: environmentVariables
             )
           case .windows:
+            guard architecture == .host else {
+              throw Error(cause: InvariantFailure(
+                """
+                Invariant failure: Runner received device with unsupported \
+                architecture '\(architecture.rawValue)' (host architecture: \
+                \(BuildArchitecture.host.rawValue))
+                """
+              ))
+            }
             try await runWindowsAppOnHost(
               bundlerOutput: bundlerOutput,
               arguments: arguments,

--- a/Sources/SwiftBundler/Bundler/SimulatorManager/Simulator.swift
+++ b/Sources/SwiftBundler/Bundler/SimulatorManager/Simulator.swift
@@ -1,6 +1,6 @@
 import Foundation
 
-struct Simulator {
+struct Simulator: Sendable, Equatable {
   var id: String
   var name: String
   var isAvailable: Bool
@@ -14,7 +14,8 @@ struct Simulator {
       id: id,
       status: isAvailable
         ? (isBooted ? .available : .summonable)
-        : .unavailable(message: "unavailable")
+        : .unavailable(message: "unavailable"),
+      architecture: .host
     )
   }
 }

--- a/Sources/SwiftBundler/Bundler/SwiftPackageManager/ApplePlatform.swift
+++ b/Sources/SwiftBundler/Bundler/SwiftPackageManager/ApplePlatform.swift
@@ -40,10 +40,24 @@ enum ApplePlatform: String, CaseIterable {
     }
   }
 
-  /// Whether the platform uses the host architecture or not. Simulators
-  /// and Mac Catalyst use the host architecture.
-  var usesHostArchitecture: Bool {
-    isSimulator || self == .macCatalyst
+  /// Architectures supported when targeting the platform.
+  var supportedArchitectures: [BuildArchitecture] {
+    switch partitioned {
+      case .macOS, .macCatalyst, .other(.simulator(_)):
+        [.arm64, .x86_64]
+      case .other(.physical(_)):
+        [.arm64]
+    }
+  }
+
+  /// Gets the default architecture to use when building for the platform.
+  func defaultTargetArchitecture(hostArchitecture: BuildArchitecture) -> BuildArchitecture {
+    switch partitioned {
+      case .macOS, .macCatalyst, .other(.simulator(_)):
+        hostArchitecture
+      case .other(.physical(_)):
+        .arm64
+    }
   }
 
   /// Whether the platform is a simulator or not.

--- a/Sources/SwiftBundler/Bundler/SwiftPackageManager/BuildArchitecture.swift
+++ b/Sources/SwiftBundler/Bundler/SwiftPackageManager/BuildArchitecture.swift
@@ -7,9 +7,9 @@ enum BuildArchitecture: String, CaseIterable, ExpressibleByArgument {
   case arm64
 
   #if arch(x86_64)
-    static let current: BuildArchitecture = .x86_64
+    static let host: BuildArchitecture = .x86_64
   #elseif arch(arm64)
-    static let current: BuildArchitecture = .arm64
+    static let host: BuildArchitecture = .arm64
   #endif
 
   var defaultValueDescription: String {

--- a/Sources/SwiftBundler/Bundler/SwiftPackageManager/Platform.swift
+++ b/Sources/SwiftBundler/Bundler/SwiftPackageManager/Platform.swift
@@ -72,6 +72,26 @@ enum Platform: String, CaseIterable {
     }
   }
 
+  /// Architectures supported when targeting the platform.
+  var supportedArchitectures: [BuildArchitecture] {
+    switch partitioned {
+      case .apple(let applePlatform):
+        applePlatform.supportedArchitectures
+      case .linux, .windows:
+        [.x86_64, .arm64]
+    }
+  }
+
+  /// Gets the default architecture to use when building for the platform.
+  func defaultTargetArchitecture(hostArchitecture: BuildArchitecture) -> BuildArchitecture {
+    switch partitioned {
+      case .apple(let applePlatform):
+        applePlatform.defaultTargetArchitecture(hostArchitecture: hostArchitecture)
+      case .linux, .windows:
+        hostArchitecture
+    }
+  }
+
   /// The platform's name.
   var name: String {
     return rawValue
@@ -113,17 +133,6 @@ enum Platform: String, CaseIterable {
         return "linux"
       case .windows:
         return "windows"
-    }
-  }
-
-  /// Whether the platform uses the host architecture or not. Simulators
-  /// and Mac Catalyst use the host architecture.
-  var usesHostArchitecture: Bool {
-    switch partitioned {
-      case .apple(let applePlatform):
-        applePlatform.usesHostArchitecture
-      case .linux, .windows:
-        true
     }
   }
 

--- a/Sources/SwiftBundler/Bundler/SwiftPackageManager/SwiftPackageManager.swift
+++ b/Sources/SwiftBundler/Bundler/SwiftPackageManager/SwiftPackageManager.swift
@@ -229,10 +229,15 @@ enum SwiftPackageManager {
         guard let platformVersion = buildContext.genericContext.platformVersion else {
           throw Error(.missingDarwinPlatformVersion(platform))
         }
-        let hostArchitecture = BuildArchitecture.current
 
+        let architectures = buildContext.genericContext.architectures
+        guard architectures.count == 1 else {
+          throw Error(.swiftPMDoesntSupportUniversalBuildsForPlatform(platform, architectures))
+        }
+
+        let architecture = architectures[0]
         let targetTriple = platform.targetTriple(
-          withArchitecture: platform.usesHostArchitecture ? hostArchitecture : .arm64,
+          withArchitecture: architecture,
           andPlatformVersion: platformVersion
         )
 

--- a/Sources/SwiftBundler/Bundler/SwiftPackageManager/SwiftPackageManagerError.swift
+++ b/Sources/SwiftBundler/Bundler/SwiftPackageManager/SwiftPackageManagerError.swift
@@ -22,6 +22,7 @@ extension SwiftPackageManager {
     case missingDarwinPlatformVersion(Platform)
     case failedToGetToolsVersion
     case invalidToolsVersion(String)
+    case swiftPMDoesntSupportUniversalBuildsForPlatform(Platform, [BuildArchitecture])
 
     var userFriendlyMessage: String {
       switch self {
@@ -63,6 +64,15 @@ extension SwiftPackageManager {
           return "Failed to get Swift package manifest tools version"
         case .invalidToolsVersion(let version):
           return "Invalid Swift tools version '\(version)' (expected a semantic version)"
+        case .swiftPMDoesntSupportUniversalBuildsForPlatform(let platform, let architectures):
+          // This should only be possible if the user has provided '--no-xcodebuild',
+          // so the error message should be pretty clear to users what they've done.
+          let clause = platform.isSimulator ? " without xcodebuild" : ""
+          return """
+            Swift Bundler cannot perform universal builds targeting \
+            \(platform.displayName)\(clause); multi-architecture build requested \
+            for \(architectures.map(\.rawValue).joinedGrammatically())
+            """
       }
     }
   }

--- a/Sources/SwiftBundler/Bundler/Xcodebuild/Xcodebuild.swift
+++ b/Sources/SwiftBundler/Bundler/Xcodebuild/Xcodebuild.swift
@@ -10,10 +10,12 @@ enum Xcodebuild {
   /// - Parameters:
   ///   - product: The product to build.
   ///   - buildContext: The context to build in.
+  ///   - destination: The destination to build for.
   /// - Returns: If an error occurs, returns a failure.
   static func build(
     product: String,
-    buildContext: SwiftPackageManager.BuildContext
+    buildContext: SwiftPackageManager.BuildContext,
+    destination: Device?
   ) async throws(Error) {
     guard let applePlatform = buildContext.genericContext.platform.asApplePlatform else {
       throw Error(.unsupportedPlatform(buildContext.genericContext.platform))
@@ -86,25 +88,13 @@ enum Xcodebuild {
     }
 
     let context = buildContext.genericContext
-    let platform = context.platform
-    let archString = context.architectures
-      .map(\.rawValue)
-      .joined(separator: "_")
 
-    var destinationString: String
-    if [.macOS, .macCatalyst].contains(platform) {
-      if context.architectures.count == 1 {
-        destinationString = "platform=macOS,arch=\(archString)"
-      } else {
-        destinationString = "generic/platform=macOS"
-      }
-    } else {
-      destinationString = "generic/platform=\(applePlatform.xcodeDestinationName)"
-    }
-    if let variant = platform.asApplePlatform?.xcodeDestinationVariant {
-      destinationString += ",variant=\(variant)"
-    }
-    let destinationArguments = ["-destination", destinationString]
+    let destinationSpecifier = try await computeDestinationSpecifier(
+      platform: applePlatform,
+      architectures: context.architectures,
+      destination: destination
+    )
+    let destinationArguments = ["-destination", destinationSpecifier]
 
     let metadataArguments: [String]
     if let compiledMetadata = buildContext.compiledMetadata {
@@ -115,6 +105,9 @@ enum Xcodebuild {
       metadataArguments = []
     }
 
+    let archString = context.architectures
+      .map(\.rawValue)
+      .joined(separator: "_")
     let suffix = context.platform.xcodeProductDirectorySuffix ?? context.platform.rawValue
     process = Process.create(
       "xcodebuild",
@@ -161,6 +154,120 @@ enum Xcodebuild {
         cause: error
       )
     }
+  }
+
+  /// Computes the xcodebuild destination specifier required to perform a build
+  /// with the given target parameters.
+  static func computeDestinationSpecifier(
+    platform: ApplePlatform,
+    architectures: [BuildArchitecture],
+    destination: Device?
+  ) async throws(Error) -> String {
+    // NOTE: The reason that this code is complicated is that xcodebuild doesn't let
+    //   us build without a destination, and destinations are either specific
+    //   devices/simulators (single architecture by definition), or generic destinations
+    //   (which are universal builds targeting all architectures supported by the target
+    //   platform). We aren't able to specify architectures directly because xcodebuild
+    //   doesn't allow `-arch` with `-destination`, and can't build SwiftPM packages without
+    //   `-destination`. Kinda sucks...
+
+    let supportedArchitectures = platform.supportedArchitectures
+    for architecture in architectures {
+      guard supportedArchitectures.contains(architecture) else {
+        throw Error(.unsupportedArchitecture(platform, architecture))
+      }
+    }
+
+    var destinationString: String
+    let platformName = platform.xcodeDestinationName
+    if architectures.count == 1 {
+      // 1. If only one architecture, perform a single architecture build
+      let architecture = architectures[0]
+      if supportedArchitectures.count == 1 {
+        // 1.1. If the target platform only supports one architecture, then
+        //   we can use a generic destination (which is ideal because it
+        //   means that you can still build even if the target device is
+        //   unavailable due to being unlocked etc).
+        destinationString = "generic/platform=\(platformName)"
+      } else if let destinationId = destination?.id {
+        // 1.2. If we have a device then that's easy
+        destinationString = "id=\(destinationId),arch=\(architecture)"
+      } else if [.macOS, .macCatalyst].contains(platform) {
+        // TODO: Does this work when targeting a non-host architecture?
+        // 1.3. If we're on macOS then it's also easy
+        destinationString = "platform=\(platformName),arch=\(architecture)"
+      } else {
+        if platform.isSimulator {
+          // 1.4.1. If we're targeting a simulator platform it gets annoying.
+          //   xcodebuild's generic simulator destinations perform universal
+          //   builds and don't let us override the architecture set, so we
+          //   have to select a target simulator.
+          let simulators = try await Error.catch {
+            try await SimulatorManager.listAvailableSimulators()
+          }
+
+          let matchingSimulators = simulators.filter { simulator in
+            // os os os, oi oi oi
+            simulator.os.os == platform.os
+          }
+
+          guard let simulator = matchingSimulators.first else {
+            throw Error(.failedToLocateSuitableDestinationSimulator(
+              simulators,
+              platform.os,
+              architecture
+            ))
+          }
+
+          destinationString = "id=\(simulator.id),arch=\(architecture)"
+        } else {
+          // 1.4.2. If we're targeting a physical non-macOS Apple device, then we
+          //   assume that there's only a single supported architecture, so we should
+          //   have reached 1.1. Therefore reaching this is an error.
+          throw Error(cause: InvariantFailure(
+            """
+            Invariant failure: Swift Bundler assumes that non-macOS Apple platforms \
+            each support only a single architecture (when not targeting a simulator). \
+            Platform '\(platform)' supports '\(supportedArchitectures)'
+            """
+          ))
+        }
+      }
+    } else if architectures == supportedArchitectures {
+      // 2. If the set of architectures matches our target platform's supported
+      //   architectures, then we can simply use a generic destination to perform
+      //   a universal build.
+      switch destination {
+        case .host, .macCatalyst, nil:
+          break
+        case .connected:
+          throw Error(.universalBuildIncompatibleWithConcreteDestination)
+      }
+
+      destinationString = "generic/platform=\(platformName)"
+    } else {
+      // 3. xcodebuild doesn't let us specify arbitrary sets of architectures when
+      //   building Swift packages, but luckily we don't have to because this should
+      //   be unreachable (unless a future Apple platform breaks out assumptions about
+      //   Apple platform architecture support).
+      throw Error(cause: InvariantFailure(
+        """
+        Invariant failure: xcodebuild can either build for a single architecture, or for all \
+        supported architectures, but not in between; selected architectures were \
+        \(architectures), supported architectures are '\(supportedArchitectures)'. \
+        This is an invariant failure; we generally assume that each Apple platform \
+        supports a maximum of 2 architectures. Please open an issue at \
+        \(SwiftBundler.gitURL.absoluteString)/issues/new
+        """
+      ))
+    }
+    
+    // Don't forget the variant!
+    if let variant = platform.xcodeDestinationVariant {
+      destinationString += ",variant=\(variant)"
+    }
+
+    return destinationString
   }
 
   /// Whether or not the bundle command utilizes xcodebuild instead of swiftpm.

--- a/Sources/SwiftBundler/Bundler/Xcodebuild/XcodebuildError.swift
+++ b/Sources/SwiftBundler/Bundler/Xcodebuild/XcodebuildError.swift
@@ -9,6 +9,13 @@ extension Xcodebuild {
     case failedToRunXcodebuild(command: String)
     case unsupportedPlatform(_ platform: Platform)
     case failedToMoveInterferingScheme(URL, destination: URL)
+    case unsupportedArchitecture(ApplePlatform, BuildArchitecture)
+    case universalBuildIncompatibleWithConcreteDestination
+    case failedToLocateSuitableDestinationSimulator(
+      [Simulator],
+      AppleOS,
+      BuildArchitecture
+    )
 
     var userFriendlyMessage: String {
       switch self {
@@ -25,6 +32,29 @@ extension Xcodebuild {
             Failed to temporarily relocate Xcode scheme at '\(relativePath)' which \
             would otherwise interfere with the build process. Move it manually \
             and try again.
+            """
+        case .unsupportedArchitecture(let platform, let architecture):
+          return """
+            Architecture '\(architecture.rawValue)' isn't supported when targeting \
+            '\(platform.platform.displayName)'
+            """
+        case .universalBuildIncompatibleWithConcreteDestination:
+          return """
+            A universal build was requested, but a destination device/simulator was
+            also provided. Cannot build for multiple architectures when targeting a
+            specific device.
+            """
+        case .failedToLocateSuitableDestinationSimulator(
+          _,
+          let os,
+          let architecture
+        ):
+          return """
+            Failed to locate suitable destination simulator with os \
+            '\(os.name)' and architecture '\(architecture.rawValue)'. \
+            xcodebuild requires that Swift Bundler provides a destination simulator \
+            for single-architecture simulator builds. Run 'swift bundler simulators \
+            list' to see available devices.
             """
       }
     }

--- a/Sources/SwiftBundler/Commands/BundleArguments.swift
+++ b/Sources/SwiftBundler/Commands/BundleArguments.swift
@@ -75,7 +75,7 @@ struct BundleArguments: ParsableArguments {
     parsing: .singleValue,
     help: {
       let possibleValues = BuildArchitecture.possibleValuesDescription
-      let defaultValue = BuildArchitecture.current.rawValue
+      let defaultValue = BuildArchitecture.host.rawValue
       return "The architectures to build for \(possibleValues). (default: [\(defaultValue)])"
     }(),
     transform: { string in

--- a/Sources/SwiftBundler/Commands/BundleCommand.swift
+++ b/Sources/SwiftBundler/Commands/BundleCommand.swift
@@ -90,6 +90,18 @@ struct BundleCommand: ErrorHandledCommand {
       }
     #endif
 
+    for architecture in arguments.architectures {
+      guard platform.supportedArchitectures.contains(architecture) else {
+        log.error(
+          """
+          Architecture '\(architecture.rawValue)' is not supported when targeting \
+          '\(platform.displayName)'
+          """
+        )
+        return false
+      }
+    }
+
     if HostPlatform.hostPlatform != .macOS && platform != HostPlatform.hostPlatform.platform {
       let hostPlatform = HostPlatform.hostPlatform.platform.displayName
       log.error("'--platform \(platform)' is not supported on \(hostPlatform)")
@@ -128,25 +140,54 @@ struct BundleCommand: ErrorHandledCommand {
       return false
     }
 
+    if [.macOS, .macCatalyst].contains(platform)
+      && arguments.architectures.count == 1
+      && arguments.architectures[0] != .host
+    {
+      if arguments.xcodebuild {
+        // It throws the following error
+        // could not find module 'MacroToolkit' for target 'arm64-apple-macos';
+        // found: x86_64-apple-macos, at: <scratch>/Build/Products/Debug/
+        // MacroToolkit.swiftmodule
+        log.warning(
+          """
+          Cross compilation for macOS with '--xcodebuild' often fails due to a \
+          bug in xcodebuild's macro handling. Perform a '--universal' build instead \
+          or omit '--xcodebuild'.
+          """
+        )
+      } else {
+        log.warning(
+          """
+          Metadata insertion is broken when cross-compiling for macOS. Build with \
+          '--universal --xcodebuild' instead to work around this issue if you \
+          rely on metadata.
+          """
+        )
+      }
+    }
+
     // macOS-only arguments
     #if os(macOS)
-      if platform.isApplePlatform && ![.macOS, .macCatalyst].contains(platform) {
-        if arguments.universal {
-          log.error(
-            """
-            '--universal' is not compatible with '--platform \
-            \(platform.rawValue)'
-            """
-          )
-          return false
-        }
+      if (arguments.universal || arguments.architectures.count > 1)
+        && (!arguments.xcodebuild || arguments.noXcodebuild)
+      {
+        log.warning(
+          """
+          SwiftPM has multiple bugs related to universal builds. If this build fails, \
+          try building with '--xcodebuild' as a workaround.
+          """
+        )
+      }
 
-        if !arguments.architectures.isEmpty {
-          log.error(
-            "'--arch' is not compatible with '--platform \(platform.rawValue)'"
-          )
-          return false
-        }
+      if case .other(.physical) = platform.asApplePlatform?.partitioned, arguments.universal {
+        log.error(
+          """
+          '--universal' is not compatible with '--platform \
+          \(platform.rawValue)'
+          """
+        )
+        return false
       }
 
       if platform != .macOS && arguments.standAlone {
@@ -377,9 +418,11 @@ struct BundleCommand: ErrorHandledCommand {
       let hostPlatform = HostPlatform.hostPlatform
       switch platform {
         case .none, hostPlatform.platform:
-          return Device.host(hostPlatform)
+          // FIXME: Resolve architecture correctly here when cross compiling
+          return Device.host(hostPlatform, .host)
         case .macCatalyst:
-          return Device.macCatalyst
+          // FIXME: Resolve architecture correctly here when cross compiling
+          return Device.macCatalyst(.host)
         case .some(let platform) where platform.isSimulator:
           let matchedSimulators = try await RichError<SwiftBundlerError>.catch {
             try await SimulatorManager.listAvailableSimulators()
@@ -438,22 +481,13 @@ struct BundleCommand: ErrorHandledCommand {
 
   func getArchitectures(platform: Platform) -> [BuildArchitecture] {
     let architectures: [BuildArchitecture]
-    switch platform {
-      case .macOS, .macCatalyst:
-        if arguments.universal {
-          architectures = [.arm64, .x86_64]
-        } else {
-          architectures =
-            !arguments.architectures.isEmpty
-            ? arguments.architectures
-            : [BuildArchitecture.current]
-        }
-      case .iOS, .visionOS, .tvOS:
-        architectures = [.arm64]
-      case .linux, .windows, .iOSSimulator, .visionOSSimulator, .tvOSSimulator:
-        architectures = [BuildArchitecture.current]
+    if arguments.universal {
+      architectures = platform.supportedArchitectures
+    } else if !arguments.architectures.isEmpty {
+      architectures = arguments.architectures
+    } else {
+      architectures = [platform.defaultTargetArchitecture(hostArchitecture: .host)]
     }
-
     return architectures
   }
 
@@ -726,7 +760,8 @@ struct BundleCommand: ErrorHandledCommand {
           if isUsingXcodebuild {
             try await Xcodebuild.build(
               product: appConfiguration.product,
-              buildContext: buildContext
+              buildContext: buildContext,
+              destination: resolvedDevice
             )
           } else {
             try await SwiftPackageManager.build(

--- a/Sources/SwiftBundler/Utility/InvariantFailure.swift
+++ b/Sources/SwiftBundler/Utility/InvariantFailure.swift
@@ -1,0 +1,13 @@
+import ErrorKit
+
+/// An error representing a failed invariant. We use this to keep invaraint
+/// failures separate from errors. Invariant failures are generally a lot less
+/// likely to be presented to users, so making an additional error enum case
+/// for each invariant that we'd like to enforce would be a wast of time.
+struct InvariantFailure: Throwable {
+  var userFriendlyMessage: String
+
+  init(_ message: String) {
+    userFriendlyMessage = message
+  }
+}

